specs/core/0_beacon-chain.md

```diff
@@ -1592,41 +1598,38 @@ _Note_: If there are skipped slots between a block and its parent block, run the
 
 ### Per-slot processing
 
-Below are the processing steps that happen at every slot.
-
-#### Slot
+Below are the processing steps that happen at every `slot >= GENESIS_SLOT`.
 
+* Set `state.latest_state_roots[state.slot % SLOTS_PER_BATCHING] = hash_tree_root(state)`.
+* Set `state.latest_block_roots[state.slot % SLOTS_PER_BATCHING] = get_block_root(state, state.slot - 1)`.
 * Set `state.slot += 1`.
-
-#### Block roots
-
-* Let `previous_block_root` be the `hash_tree_root` of the previous beacon block processed in the chain.
-* Set `state.latest_block_roots[(state.slot - 1) % LATEST_BLOCK_ROOTS_LENGTH] = previous_block_root`.
-* If `state.slot % LATEST_BLOCK_ROOTS_LENGTH == 0` append `merkle_root(state.latest_block_roots)` to `state.batched_block_roots`.
+* If `state.slot % SLOTS_PER_BATCHING == 0` append `merkle_root(state.latest_block_roots + state.latest_state_roots)` to `state.historical_batchings`.
 
 ### Per-block processing
 
 Below are the processing steps that happen at every `block`.
 
-#### Slot
-
-* Verify that `block.slot == state.slot`.
-
-#### Block signature
+#### Block header
 
+* Verify that `block.header.slot == state.slot`.
+* Verify that `block.header.block_body_root == hash_tree_root(block.body)`.
+* Set `state.latest_partial_block.header.state_root = get_state_root(state, state.slot - 1)`.
```

Comment:
This section is pretty opaque. The block roots that go into state when there are empty slots involved having a modified state root? And the block header is supposed to point to this modified block instead of it's actual parent?

The block needs to point to another real block, otherwise the client cannot figure out where the state transition to the block is from. It would have to state transition from all possible blocks to the block's prev-slot to try to find the possible parent root.

I might be missing something here. can you write out in a few plain english steps what we are trying to accomplish here?

Reply:
The idea is that `state.latest_partial_block` contains the latest block with `header.state_root` zeroed out (to avoid the recursion). On line 1616 we reconstruct the (full) previous block with `header.state_root` filled in, and on line 1617 we check that the `previous_block_root` matches. I'll create the `previous_block` variable to clear things up.